editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -0,0 +1,359 @@
+            updateTopLevelElements(
+              filename,
+              (changeEntry.newValue as CollabTextFileTopLevelElements).toArray(),
+            ),
+          )
+        }
+        break
+      case 'delete':
+        actions.push(deleteFile(filename))
+        break
+      default:
+        throw new Error(`Unhandled change entry action: ${changeEntry.action}`)
+    }
+    entriesIteratorResult = entriesIterator.next()
+  }
+  dispatch(actions)
+}
+
+function updateTopLevelElementsOfFile(
+  session: CollaborativeEditingSupportSession,
+  filePath: string,
+  changeEvent: Y.YEvent<any>,
+  dispatch: EditorDispatch,
+): void {
+  const file = session.projectContents.get(filePath)
+  if (file != null) {
+    const oldTopLevelElements = file.get('topLevelElements') as CollabTextFileTopLevelElements
+    let newTopLevelElements: Array<TopLevelElement> = []
+    let readIndex = 0
+    for (const delta of changeEvent.delta) {
+      if (delta.retain != undefined) {
+        const elementsToPush = oldTopLevelElements.slice(readIndex, readIndex + delta.retain)
+        newTopLevelElements.push(...elementsToPush)
+        readIndex += delta.retain
+      }
+      if (delta.insert != null && Array.isArray(delta.insert)) {
+        newTopLevelElements.push(...delta.insert)
+      }
+      if (delta.delete != undefined) {
+        readIndex += delta.delete
+      }
+    }
+
+    if (readIndex < oldTopLevelElements.length) {
+      // There is an implicit retain for the remainder of the items
+      const elementsToPush = oldTopLevelElements.slice(readIndex)
+      newTopLevelElements.push(...elementsToPush)
+    }
+
+    dispatch([updateTopLevelElements(filePath, newTopLevelElements)])
+  }
+}
+
+function arrayChanges(updatesAt: Array<number>, deleteFrom: number | null): ArrayChanges {
+  return {
+    updatesAt: updatesAt,
+    deleteFrom: deleteFrom,
+  }
+}
+
+export function calculateArrayChanges<T>(
+  from: Array<T>,
+  into: Y.Array<T>,
+  equals: (from: T, into: T) => boolean = (fromToCheck, intoToCheck) => fromToCheck === intoToCheck,
+): ArrayChanges {
+  let updatesAt: Array<number> = []
+  for (let index: number = 0; index < Math.max(from.length, into.length); index++) {
+    if (!equals(from[index], into.get(index))) {
+      updatesAt.push(index)
+    }
+  }
+
+  let deleteFrom: number | null = null
+  if (into.length > from.length) {
+    deleteFrom = into.length
+  }
+
+  return arrayChanges(updatesAt, deleteFrom)
+}
+
+function synchroniseParseSuccessToCollabFile(
+  success: ParseSuccess,
+  collabFile: CollabTextFile,
+): void {
+  const collabFileTopLevelElements = collabFile.get('topLevelElements')
+  if (collabFileTopLevelElements === 'TEXT_FILE' || collabFileTopLevelElements === undefined) {
+    throw new Error('Invalid value for topLevelElements.')
+  } else {
+    const changes = calculateArrayChanges(
+      success.topLevelElements,
+      collabFileTopLevelElements,
+      // Fix this shitshow up
```

Comment:
👀 should this be fixed then?

Reply:
Hahaha this was my original comment about the equality function below, but I dunno, maybe it's not that bad anymore 🤷

Reply:
Comment removed in 6d755a897.